### PR TITLE
ls: Added cmd to check when user was last seen.

### PR DESCRIPTION
**Pull request by JakubNvk**
Issue: #95 
To: @mrshu 
Cc: @Adman @pepol 

Signed-off-by: Jakub Novak <jakub.novak@gmail.com>

### CODE REVIEW

File: last_seen.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+class LastSeen(BotPlugin):
+    def setup(self, *args, **kwargs):
+        self.seen = self.open_storage('last_seen')
+
+    @event
+    def seen_evt(self, event):
+        events = ['quit', 'part', 'kick', 'join', 'rename', 'topic']
+        if event.event_type in events:
+            nick = event.meta['nick']
+            time = strftime("%d.%m.%Y", gmtime())
```
Comment (Adman) [inline last_seen.py]: we probably want to know hours/minutes, don't we?

You could save the time of last activity, and then in last_seen cmd return something like "2 hours ago"

File: last_seen.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            self.seen[nick] = time
```
Comment (mrshu) [inline last_seen.py]: @JakubNvk personally, I'd also like to know the event (was it a message? Did the user leave the channel?)

File: last_seen.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+    @cmd
+    def last_seen(self, event):
+        print strftime("%d %m %Y", gmtime())
```
Comment (mrshu) [inline last_seen.py]: @JakubNvk you probably didn't want to include this line.

**Comment by mrshu** [comment]
@JakubNvk apart from those comments by me and @Adman I think it looks pretty good.

File: last_seen.py (hunk 2 above)
Comment (JakubNvk) [inline last_seen.py]: @mrshu well if you dont see the user there, it's safe to say he left the channel :)

File: last_seen.py (hunk 3 above)
Comment (JakubNvk) [inline last_seen.py]: @mrshu you're right.

File: last_seen.py (hunk 1 above)
Comment (JakubNvk) [inline last_seen.py]: @Adman sure, I can add time aswell

File: last_seen.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    def setup(self, *args, **kwargs):
+        self.seen = self.open_storage('last_seen')
+
+    @event
+    def seen_evt(self, event):
+        events = ['quit', 'part', 'kick', 'join', 'rename', 'topic']
+        if event.event_type in events:
+            nick = event.meta['nick']
+            time = strftime("%H:%M %d.%m.%Y", gmtime())
+            self.seen[nick] = time
```
Comment (mrshu) [inline last_seen.py]: @JakubNvk I still believe you should remember the `(time, event.event_type)` touple and then use the second part in `last_seen` to let everyone know what was the last user's action.

Comment (Adman) [inline last_seen.py]: @JakubNvk and also, don't strftime here. Save raw time instead.

Comment (JakubNvk) [inline last_seen.py]: @mrshu but what is it good for? If you don't see the user there, he's left. 99% of the time the event will be `part`, `quit` or `kick`.
@Adman why is that?

Comment (mrshu) [inline last_seen.py]: @JakubNvk please do not let your biased view of a small channel fool you. Many users use bouncers and other methods of staying always online. In a huge channel where there are many simultanious discussions one can get lost pretty easily. Thus, it is indeed necessary in my opinion to provide the user with as much context as possible and that definitely includes the event.

File: last_seen.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+from brutal.core.plugin import BotPlugin, cmd, event
+from time import strftime, gmtime
+
+class LastSeen(BotPlugin):
+    def setup(self, *args, **kwargs):
+        self.seen = self.open_storage('last_seen')
+
+    @event
+    def seen_evt(self, event):
+        events = ['quit', 'part', 'kick', 'join', 'rename', 'topic']
```
Comment (mrshu) [inline last_seen.py]: @JakubNvk why is `message` not in here? What if the user is stil connected but not active in a channel? Wouldn't you want to see his or her last activity which might easily be sending a message?

File: last_seen.py (hunk 4 above)
Comment (JakubNvk) [inline last_seen.py]: @mrshu allright, now it makes more sense to add it :)

Comment (Adman) [inline last_seen.py]: @JakubNvk Because you can do simple diff `current_time - last_seen_time` and then strftime that result

Comment (mrshu) [inline last_seen.py]: @JakubNvk regarding @Adman's note -- that will give you the ability to decide what to do with a raw date afterwards. In the current setup you can only show the string you have saved.

File: last_seen.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
+from brutal.core.plugin import BotPlugin, cmd, event
+from time import strftime, gmtime
+
+class LastSeen(BotPlugin):
```
Comment (mrshu) [inline last_seen.py]: @JakubNvk the travis tests will fail if there won't be two blank lines before a class definition. Please add a blank line here. Thanks

File: last_seen.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        self.seen = self.open_storage('last_seen')
+
+    @event
+    def seen_evt(self, event):
+        events = ['quit', 'part', 'kick', 'join', 'rename', 'topic', 'message']
+        if event.event_type in events:
+            nick = event.meta['nick']
+            time = gmtime()
+            evt_type = event.event_type
+            self.seen[nick] = {time: evt_type}
```
Comment (Adman) [inline last_seen.py]: This is not how dictionary works :)

File: last_seen.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        user = args[0]
+        seen = self.seen
+
+        if user not in seen:
+            return "This user was never here." + strftime("%H:%M %d.%m.%Y")
+
+        raw_time = seen[user].keys()[0]
+        evt = seen[user][raw_time]
+        time = strftime("%H:%M %d.%m.%Y", raw_time)
+        return "{0} was last seen: {1}({2})".format(user, time, evt)
```
Comment (Adman) [inline last_seen.py]: Wouldn't it be better to say just `x hours ago` ?

File: last_seen.py (hunk 7 above)
Comment (mrshu) [inline last_seen.py]: @adman technically it is but I doubt this is what we want

Comment (JakubNvk) [inline last_seen.py]: @mrshu and what do we want?

File: last_seen.py (hunk 8 above)
Comment (mrshu) [inline last_seen.py]: @adman suppose someone  hasn't  shown up for 20 days. I think x hours ago maes sense  only in the first 24 hours.

File: last_seen.py (hunk 7 above)
Comment (mrshu) [inline last_seen.py]: @jakubn I think a basic touple `(time, evt_type)` would be enough. You could then access  it via `self.seen[nick][0]` for instance. The 1st index would return `event_time`.

File: last_seen.py (hunk 8 above)
Comment (JakubNvk) [inline last_seen.py]: @mrshu :+1: exactly my thinking.

File: last_seen.py (hunk 7 above)
Comment (JakubNvk) [inline last_seen.py]: @mrshu I´ll go with a proper dictionary, I like it more that way.

File: last_seen.py (hunk 8 above)
Comment (Adman) [inline last_seen.py]: @mrshu Yeah, then we could show up something like `x days y hours ago`

File: last_seen.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+            evt_type = event.event_type
+            self.seen[nick] = {'time': time, 'event': evt_type}
+
+    @cmd
+    def last_seen(self, event):
+        args = event.args
+        if len(args) < 1:
+            return "Name argument missing."
+        user = args[0]
+        seen = self.seen
```
Comment (Adman) [inline last_seen.py]: What for do you create new dict with the same stuff like self.seen?

File: last_seen.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    def last_seen(self, event):
+        args = event.args
+        if len(args) < 1:
+            return "Name argument missing."
+        user = args[0]
+
+        if user not in self.seen:
+            return "This user was never here."
+
+        evt = self.seen[user][0]
```
Comment (mrshu) [inline last_seen.py]: @jakubnvk why not `evt, time = self.seen [user]`?

Comment (mrshu) [inline last_seen.py]: @jakubnvk sorry I meant `*self.seen...`

File: last_seen.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+        args = event.args
+        if len(args) < 1:
+            return "Name argument missing."
+        user = args[0]
+
+        if user not in self.seen:
+            return "This user was never here."
+
+        evt, time = self.seen[user]
+        ret_time = strftime("%H:%M %d.%m.%Y", time)
```
Comment (mrshu) [inline last_seen.py]: This is just waste of space and memory but hopefully  the interpreter  will pick it up.

File: last_seen.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+class LastSeen(BotPlugin):
+    def setup(self, *args, **kwargs):
+        self.seen = self.open_storage('last_seen')
+
+    @event
+    def seen_evt(self, event):
+        events = ['quit', 'part', 'kick', 'join', 'rename', 'topic', 'message']
+        if event.event_type in events:
+            nick = event.meta['nick']
```
Comment (mrshu) [inline last_seen.py]: Same as below, I do not see any reason for this.

File: last_seen.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+        if len(args) < 1:
+            return "Name argument missing."
+        user = args[0]
+
+        if user not in self.seen:
+            return "This user was never here."
+
+        evt, time = self.seen[user]
+        ret_time = strftime("%H:%M %d.%m.%Y", time)
+        return "{0} was last seen: {1}({2})".format(user, ret_time, evt)
```
Comment (mrshu) [inline last_seen.py]: What hppened to x hours ago?

Comment (JakubNvk) [inline last_seen.py]: @mrshu nothing, I don't feel it's necessary.

File: last_seen.py (hunk 11 above)
Comment (JakubNvk) [inline last_seen.py]: If I don't do it this way, the return just looks terrible.

Comment (mrshu) [inline last_seen.py]: @JakubNvk you could reassign the strftime-d value back to time -- it's not used anywhere else anyway.

File: last_seen.py [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    def seen_evt(self, event):
+        events = ['quit', 'part', 'kick', 'join', 'rename', 'topic', 'message']
+        if event.event_type in events:
+            nick = event.meta['nick']
+            time = gmtime()
+            evt_type = event.event_type
+            self.seen[nick] = (evt_type, time)
+
+    @cmd
+    def last_seen(self, event):
```
Comment (mrshu) [inline last_seen.py]: @JakubNvk could you please add a short help message along with an example please?

**Comment by mrshu** [comment]
@JakubNvk ok, looks pretty good to me, if @Adman is not agaist it feel free to merge this in.

**Pull request merged**